#  Отчет по практическому заданию №2 
## Постановка задачи

   Написать и отладить процедуру, вычисляющую определенный интеграл от функции одного действительного переменного по квадратурной формуле Симпсона. Подпрограмма должна обеспечивать автоматический контроль точности и деление интервала интегрирования на подынтервалы там, где это необходимо.


Формула Симпсона основана на замене подынтегральной функции f(x) на отрезке [a, b] дугой параболы, т.е. функция f(x) аппроксимируется параболой вида: $P(x)=\alpha x^2 + \beta x + \gamma$.

Разобъем отрезок $[a,b]$ на четное число равных отрезков $n = 2m$, при этом точки $x_0, x_2, x_4, ... , x_{n-2}, x_n$ - точки деления ($x_0 = a, x_n = b$). Обозначим через $x_1, x_3, x_5$, ... середины отрезков $[x_0, x_2], [x_2, x_4], [x_4, x_6]$ и т.д. Применив для каждого отрезка разбиения элементарную формулу Симпсона, получим формулу парабол.

### $\displaystyle{\int\limits_a^b f(x) dx \approx \frac{h}{3} ~ [y_0 + y_n + 2(y_{2}+y_4+...+y_{n-2})+4(y_1+y_3+...+y_{n-1})]}$ 

### Введем начальные данные

In [197]:
n=10    # кол-во интервалов разбиения 
a=-2    # нижний предел интегрирования
b=1     # верхний предел интегрирования

            # задаем функцию, y1 для подсчета производной, f(x) для нахоождения интеграла
y1=x1**6 -x1**4   
def f(x):
    return float(x)**6 -float(x)**4

In [198]:
import numpy as np
h = (b-a)/n       # интервал 
x=np.arange(a, b+h, h)    
print(x)

[-2.  -1.7 -1.4 -1.1 -0.8 -0.5 -0.2  0.1  0.4  0.7  1. ]


In [199]:
c2 = 0
c4 = 0
for i in range(1, int((n)/2)):
    c2 += f(x[2*i])   
    c4 += f(x[2*i-1])
        
c4 += f(x[n-1])

I = h/3 * (f(x[0])+f(x[n])+2*c2+4*c4)

print(I)

11.872889999999996


### Интеграл равен I = 11.87

## Поиск остаточного члена R
$ \displaystyle{R \simeq - \frac{(b-a) ~ h^4 ~ f^{IV}_{max}}{180}} $

### Поиск четвертой производной

In [200]:
from sympy import *
x1 = symbols('x')
pr4 = diff(y1,x1,4)  # взяли четвертую производную по х

print(pr4)

24*(15*x**2 - 1)


### Поиск максимума четвертой производной на интервале $[a,b]$

In [201]:
from sympy import solveset, symbols, Interval, Min, Max
x = symbols('x')

lower_bound = a
upper_bound = b

f = pr4

zeros = solveset(f, x, domain=Interval(lower_bound, upper_bound))
assert zeros.is_FiniteSet # If there are infinite solutions the next line will hang.
res_min = Min(f.subs(x, lower_bound), f.subs(x, upper_bound), *[f.subs(x, i) for i in zeros])
res_max = Max(f.subs(x, lower_bound), f.subs(x, upper_bound), *[f.subs(x, i) for i in zeros])
print(res_max)

1416


### Точность вычисления интеграла

In [202]:
R = - (b-a)/180*h**4*res_max
print(R)

-0.191160000000000


## Таким образом, $I = 11.9  \pm 0.2 $